### PR TITLE
Bugfix/ps 3151 tldk downloads slow or pause

### PR DESCRIPTION
**Pull request by stefan-baranoff**

### CODE REVIEW

File: lib/l4p/tcp_rxtx.c [hunk 1]
```diff
@@ -2466,6 +2499,30 @@ tle_tcp_stream_readv(struct tle_stream *ts, const struct iovec *iov,
 	}
 
 	tcp_rxq_consume(s, tn);
+	// If the window went from 0 to non-0, we need to send an update.
+    //
+	// There is a slight change of a race that could double-ack the same data:
+	// 1. The back end may have enqueued packets to the RX queue putting the window at 0
+	// 2. We read data here on the front end opening the window
+	// 3. The back end queues up an ack to send, beceause it got packets
```
Comment (ttosbt) [inline lib/l4p/tcp_rxtx.c]:
```suggestion
	// 3. The back end queues up an ack to send, because it got packets
```

File: lib/l4p/tcp_rxtx.c [hunk 2]
```diff
@@ -206,15 +206,16 @@ tcp_stream_adjust_tms(const struct tle_tcp_stream *s, uint32_t tms)
 }
 
 static inline void
-fill_tcph(struct rte_tcp_hdr *l4h, const struct tcb *tcb, union l4_ports port,
+fill_tcph(struct rte_tcp_hdr *l4h, const struct tle_tcp_stream *s, union l4_ports port,
 	uint32_t seq, uint8_t hlen, uint8_t flags)
 {
+	const struct tcb *tcb = &s->tcb;
```
Comment (ttosbt) [inline lib/l4p/tcp_rxtx.c:212]: I may be missing something, but you've added a new argument...then here pulled out a part of it...then never use it again.

Should this line be here?

File: lib/l4p/tcp_rxtx.c [hunk 3]
```diff
@@ -303,17 +304,18 @@ tcp_fill_mbuf(struct rte_mbuf *m, const struct tle_tcp_stream *s,
  *  - if no HW cksum offloads are enabled, calculates TCP checksum.
  */
 static inline void
-tcp_update_mbuf(struct rte_mbuf *m, uint32_t type, const struct tcb *tcb,
+tcp_update_mbuf(struct rte_mbuf *m, uint32_t type, const struct tle_tcp_stream *s,
 	uint32_t seq, uint32_t pid, uint8_t tcp_flags)
 {
+	const struct tcb *tcb = &s->tcb;
```
Comment (ttosbt) [inline lib/l4p/tcp_rxtx.c:310]: Same question as above.

File: lib/l4p/tcp_rxtx.c [hunk 4]
```diff
@@ -471,12 +482,33 @@ static inline uint32_t
...
+	// The window multiplier is 256, so the receiver can't report exactly 50,000
+	// bytes of buffer. It can report either 49,920 or 50,176. It chooses 49,920.
+	//
+	// Now there are 50,000 bytes in flight but snd.wnd is 49,920 and subtraction
+	// produces a negative (or underflowed an unsigned type to large positive) number.
+	//
+	// We need a saturating truncate, not an underflowing truncate,
+	// so use a signed value with max.
+
+	swnd = (int64_t)s->tcb.snd.wnd - (int64_t)(s->tcb.snd.nxt - s->tcb.snd.una);
```
Comment (ttosbt) [inline lib/l4p/tcp_rxtx.c:510]: Guess this is just a styling thing?  You declare the variable above but then you set it here...just make it one line?  Additionally, it doesn't look like you're changing it after creation so make it const as well?

File: lib/l4p/tcp_rxtx.c [hunk 5]
```diff
@@ -495,9 +527,8 @@ tx_nxt_data(struct tle_tcp_stream *s, uint32_t tms)
 			n = tx_data_bulk(s, &sl_una, mi, 1);
 			tn += n;
 			// of we enqueued the segment go back to normal processing
-			if (n == 1) {
+			if (n == 1)
```
Comment (ttosbt) [inline lib/l4p/tcp_rxtx.c:530]: Abandon hope all ye who enter here...haha.  Call me old fashioned, but brackets for an if block just seems more natural than not.

File: lib/l4p/tcp_rxtx.c (hunk 2 above)
Comment (stefan-baranoff) [reply lib/l4p/tcp_rxtx.c:212]: It is used; just outside of the diff window. Look at (new) lines 222, 230, and 232.

File: lib/l4p/tcp_rxtx.c (hunk 3 above)
Comment (stefan-baranoff) [reply lib/l4p/tcp_rxtx.c:310]: Same answer - used but out of diff view.

File: lib/l4p/tcp_rxtx.c (hunk 4 above)
Comment (stefan-baranoff) [reply lib/l4p/tcp_rxtx.c:510]: Yep; mostly a styling thing. Older C-style declaration separate from definition; welcome to the pain.

File: lib/l4p/tcp_rxtx.c (hunk 5 above)
Comment (stefan-baranoff) [reply lib/l4p/tcp_rxtx.c:530]: Welcome to TLDK/DPDK styling. I agree, but they generally don't.

**Pull request merged**